No views in session and error loading packages

Hello again! 👋🏻 

I'm still trying to get Gore to work, but have been unsuccessful so far. Since #1, I updated to the latest lsp4e snapshot, 0.13.9.202201201104, and to the latest gopls release, 0.7.5.

When I open a Go file, the plugin is able to find the language server executable, as it logs `GORE: Found gopls at ...` with the right path. However, immediately after it logs the following error message:
```
java.util.concurrent.CompletionException: org.eclipse.lsp4j.jsonrpc.ResponseErrorException: no views in session
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:332)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:347)
	at java.base/java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java:708)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2162)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleResponse(RemoteEndpoint.java:209)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:193)
	at org.eclipse.lsp4e.LanguageServerWrapper.lambda$4(LanguageServerWrapper.java:260)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.handleMessage(StreamMessageProducer.java:194)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:94)
	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor.run(ConcurrentMessageProcessor.java:113)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: org.eclipse.lsp4j.jsonrpc.ResponseErrorException: no views in session
	... 11 more
```

A popup then gets continuously duplicated several times per second, quickly making the IDE unresponsive and slowing down the whole system:
<img width="450" alt="Screenshot 2022-01-22 at 20 50 01" src="https://user-images.githubusercontent.com/10694593/150653510-691a5636-0100-4a50-9ee4-5c44d7af5d7f.png">

The error message is truncated, but it claims that it can't load packages as it seemingly can't find the go command. This is odd given that it's able to find gopls, and that gopls starts happily when launched from the command line.

Any ideas?

--- a/ca.mt.gore/src/ca/mt/gore/GoResolver.java
+++ b/ca.mt.gore/src/ca/mt/gore/GoResolver.java
@@ -1,0 +1,84 @@
+/*******************************************************************************
+ * Copyright (c) 2021 TODO and others.
+ *
+ * This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     TODO - initial API and implementation
+ ******************************************************************************/
+
+package ca.mt.gore;
+
+import com.google.common.base.Splitter;
+import com.google.common.base.Strings;
+import com.google.common.io.ByteStreams;
+import java.io.File;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.List;
+import java.util.function.Supplier;
+import org.eclipse.core.runtime.Platform;
+
+/**
+ * Helper class for resolving Go-related binaries.
+ */
+public class GoResolver {
+	public static String resolveGoBinary(String program) {
+		if (Platform.OS_WIN32.equals(Platform.getOS()) && !program.toLowerCase().endsWith(".exe")) {
+			program += ".exe";
+		}
+		List<Supplier<String>> pathProviders = Arrays.asList(() -> System.getenv("GOBIN"),
+				() -> System.getenv("GOPATH"), () -> run("go", "env", "GOBIN"), () -> run("go", "env", "GOPATH"),
+				() -> new File(System.getenv("HOME"), "go").toString());
+		for (Supplier<String> provider : pathProviders) {
+			String path = provider.get();
+			if (!Strings.isNullOrEmpty(path)) {
+				String fullPath = resolve(path, program);
+				if (fullPath != null) {
+					return fullPath;
+				}
+				fullPath = resolve(path, "bin" + File.separator + program);
+				if (fullPath != null) {
+					return fullPath;
+				}
+			}
+		}
+
+		return resolve(System.getenv("PATH"), program);
+	}
+
+	private static String resolve(String env, String relativeFile) {
+		if (Strings.isNullOrEmpty(env)) {
+			return null;
+		}
+		env = env.trim();
+		for (String path : Splitter.on(File.pathSeparator).split(env)) {
+			File file = new File(path, relativeFile);
+			if (file.exists() && file.canExecute()) {
+				return file.getPath();
+			}
+		}
+		return null;
+	}
+
+	private static String run(String... cmd) {
+		try {
+			Process process = new ProcessBuilder(cmd).start();
+			if (process.waitFor() != 0) {
+				return null;
+			}
+			return new String(ByteStreams.toByteArray(process.getInputStream()), StandardCharsets.UTF_8).trim();
+		} catch (IOException ex) {
+			return null;
+		} catch (InterruptedException ex) {
+			return null;
+		}
+	}
+
+}

--- a/ca.mt.gore/src/ca/mt/gore/GoServer.java
+++ b/ca.mt.gore/src/ca/mt/gore/GoServer.java
@@ -1,8 +1,6 @@
 package ca.mt.gore;
 
-import com.google.common.base.Splitter;
 import com.google.common.base.Strings;
-import com.google.common.io.ByteStreams;
 import com.google.gson.Gson;
 import com.google.gson.JsonElement;
 import com.google.gson.JsonObject;
@@ -14,12 +12,8 @@ import java.io.InputStream;
 import java.io.OutputStream;
 import java.io.Reader;
 import java.net.URI;
-import java.nio.charset.StandardCharsets;
-import java.util.ArrayList;
-import java.util.Arrays;
 import java.util.List;
 import java.util.Map;
-import java.util.function.Supplier;
 import org.eclipse.core.runtime.Platform;
 import org.eclipse.lsp4e.LanguageServerPlugin;
 import org.eclipse.lsp4e.server.ProcessOverSocketStreamConnectionProvider;
@@ -27,9 +21,38 @@ import org.eclipse.lsp4e.server.ProcessStreamConnectionProvider;
 import org.eclipse.lsp4e.server.StreamConnectionProvider;
 
 public class GoServer implements StreamConnectionProvider {
+	private final class ProcessStreamConnectionProviderExtension extends ProcessStreamConnectionProvider {
+		private ProcessStreamConnectionProviderExtension(List<String> commands, String workingDir) {
+			super(commands, workingDir);
+		}
+
+		@Override
+		protected ProcessBuilder createProcessBuilder() {
+			ProcessBuilder pb = super.createProcessBuilder();
+			addGoplsLocationToEnvironment(pb.environment());
+			return pb;
+		}
+	}
+
+	private final class ProcessOverSocketStreamConnectionProviderExtension
+			extends ProcessOverSocketStreamConnectionProvider {
+		private ProcessOverSocketStreamConnectionProviderExtension(List<String> commands, String workingDir, int port) {
+			super(commands, workingDir, port);
+		}
+
+		@Override
+		protected ProcessBuilder createProcessBuilder() {
+			ProcessBuilder pb = super.createProcessBuilder();
+			addGoplsLocationToEnvironment(pb.environment());
+			return pb;
+		}
+	}
+
 	private static final int CONNECTION_PORT = 4389;
 
 	private StreamConnectionProvider provider;
+
+	private String goplsDir;
 
 	public GoServer() {}
 
@@ -93,84 +116,33 @@ public class GoServer implements StreamConnectionProvider {
 
 	private void createProvider() throws IOException {
 		String os = Platform.getOS();
-		List<String> commands = new ArrayList<>();
 
-		String gopls = resolveGoBinary("gopls");
-		if (gopls != null) {
-			LanguageServerPlugin.logInfo("GORE: Found gopls at " + gopls);
-			commands.add(gopls);
-			commands.add("serve");
-			// commands.add("-rpc.trace");
-			if (os.equals(Platform.OS_WIN32)) {
-				commands.add("--listen=127.0.0.1:" + CONNECTION_PORT);
-			}
-		}
-
-		if (commands.isEmpty()) {
+		String gopls = GoResolver.resolveGoBinary("gopls");
+		if (gopls == null) {
 			IOException ex = new IOException("Cannot find gopls");
 			LanguageServerPlugin.logError("GORE: Could not find gopls or bingo", ex);
 			throw ex;
 		}
 
+		LanguageServerPlugin.logInfo("GORE: Found gopls at " + gopls);
+		goplsDir = gopls.substring(0, gopls.lastIndexOf(File.separator));
+
+		List<String> commands = List.of(gopls, "serve"); // "-rpc.trace"
 		File workingDir = new File(".");
 		if (os.equals(Platform.OS_WIN32)) {
-			provider = new ProcessOverSocketStreamConnectionProvider(commands, workingDir.toString(),
-					CONNECTION_PORT) {};
+			commands.add("--listen=127.0.0.1:" + CONNECTION_PORT);
+			provider = new ProcessOverSocketStreamConnectionProviderExtension(commands, workingDir.toString(),
+					CONNECTION_PORT);
 		} else {
-			provider = new ProcessStreamConnectionProvider(commands, workingDir.toString()) {};
+			provider = new ProcessStreamConnectionProviderExtension(commands, workingDir.toString());
 		}
 	}
 
-	private static String resolveGoBinary(String program) {
-		if (Platform.OS_WIN32.equals(Platform.getOS()) && !program.toLowerCase().endsWith(".exe")) {
-			program += ".exe";
-		}
-		List<Supplier<String>> pathProviders = Arrays.asList(() -> System.getenv("GOBIN"),
-				() -> System.getenv("GOPATH"), () -> run("go", "env", "GOBIN"), () -> run("go", "env", "GOPATH"),
-				() -> new File(System.getenv("HOME"), "go").toString());
-		for (Supplier<String> provider : pathProviders) {
-			String path = provider.get();
-			if (!Strings.isNullOrEmpty(path)) {
-				String fullPath = resolve(path, program);
-				if (fullPath != null) {
-					return fullPath;
-				}
-				fullPath = resolve(path, "bin" + File.separator + program);
-				if (fullPath != null) {
-					return fullPath;
-				}
-			}
-		}
-
-		return resolve(System.getenv("PATH"), program);
-	}
-
-	private static String run(String... cmd) {
-		try {
-			Process process = new ProcessBuilder(cmd).start();
-			if (process.waitFor() != 0) {
-				return null;
-			}
-			return new String(ByteStreams.toByteArray(process.getInputStream()), StandardCharsets.UTF_8).trim();
-		} catch (IOException ex) {
-			return null;
-		} catch (InterruptedException ex) {
-			return null;
-		}
-	}
-
-	private static String resolve(String env, String relativeFile) {
-		if (Strings.isNullOrEmpty(env)) {
-			return null;
-		}
-		env = env.trim();
-		for (String path : Splitter.on(File.pathSeparator).split(env)) {
-			File file = new File(path, relativeFile);
-			if (file.exists() && file.canExecute()) {
-				return file.getPath();
-			}
-		}
-		return null;
+	/**
+	 * Update the gopls environment by adding the gopls location to the PATH.
+	 */
+	public void addGoplsLocationToEnvironment(Map<String, String> environment) {
+		environment.merge("PATH", goplsDir, (pathValue, update) -> pathValue + File.pathSeparator + update);
 	}
 
 	@Override

